Error reporting

chmeliik opened the pull request.

MartinBasti commented on operatorcourier/errors.py:
```diff
@@ -0,0 +1,46 @@
+    """Invalid bundle (e.g. missing CSV/CRD/Package).
+    Contains the info collected during validation of the bundle.
+    """
+    def __init__(self, msg, validation_dict):
```
I'd call this `validation_results`, we shouldn't name variables with types in python

MartinBasti commented on operatorcourier/errors.py:
```diff
@@ -0,0 +1,46 @@
+    def __init__(self, msg, code=500):
+        """
+        :param msg: the message of this exception
+        :param code: http status code returned by request
```
s/request/response/ maybe ???

MartinBasti commented on tests/test_api.py:
```diff
@@ -37,3 +38,29 @@ def test_make_bundle_with_yaml_list(yaml_files, expected):
+        with open(file, "r") as yaml_file:
+            yamls.append(yaml_file.read())
+
+    with pytest.raises(OpCourierBadBundle) as e_info:
```
why `e_info` and not just `e` or `exc`?

chmeliik replied:
`pytest.raises()` wraps the actual exception in an `ExceptionInfo` object, hence `e_info`. I'm down to change it though if that makes it more readable (tomorrow, that is :smile: )

MartinBasti replied:
Ah I didn't realize that, it is ok then :D

MartinBasti commented on operatorcourier/errors.py:
```diff
@@ -0,0 +1,46 @@
+"""
+
+
+class OpCourierError(ValueError):
```
Could we add unit test that `OpCourierError` (and children) are subclass of `ValueError` to make sure we will keep this BW compatibility. Unless @kevinrizza agree to break compatibility

kevinrizza replied:
Maintaining backwards compatibility is a major goal in the short term, since we have only recently released. Making breaking changes should be reserved for a specific release goal, which this pr is not aiming for. Please maintain backwards compatibility.

MartinBasti commented:
@chmeliik Good job, I had just minor comments! Let's wait for owners if they agree with such changes :-)

MartinBasti commented:
Maybe public API methods should also have documented which exceptions they raise.

kevinrizza commented on operatorcourier/errors.py:
```diff
@@ -0,0 +1,46 @@
+
+class OpCourierQuayError(OpCourierError):
+    """An error occurred while pushing bundle to Quay.io"""
+    def __init__(self, msg, code=500):
```
Why do we have a default here? If there is an error returned from the server, we should always have an error code (even if the error is that the server is not available -- which is not 500).

chmeliik replied:
That's fair. Will remove default value.

MartinBasti replied:
I was thinking about this, I'd really like to get content from quay what is wrong with error code and quay error JSON if possible.

We may actually split this into more exceptions:
* OpCourierQuayError -- generic quay error
* OpCourierQuayCommunicationError(OpCourierQuayError) - random error that happened during communication
*OpCourierQuayErrorResponse(OpCourierQuayError) -- error returned from quay with quay details, basically catching HTTPException from requests.

MartinBasti replied:
@chmeliik This got +1 from Kevin

chmeliik replied:
Understood, will get on it tomorrow

chmeliik commented:
Updated PR, cc @kevinrizza  @MartinBasti  @awgreene

kevinrizza commented:
@chmeliik Can you please squash your commits?

chmeliik commented:
@kevinrizza Can do, do you want all the changes in a single commit?

kevinrizza commented:
> @kevinrizza Can do, do you want all the changes in a single commit?

@chmeliik Yes please.

MartinBasti commented on tests/test_errors.py:
```diff
@@ -0,0 +1,19 @@
+    OpCourierQuayError,
+    OpCourierError
+])
+def test_exception_backwards_compatibility(op_courier_exception):
```
Please add docstring that we are testing BW compatiblity

chmeliik replied:
Done

chmeliik commented:
Added specific Quay errors, cc @kevinrizza @MartinBasti @awgreene

kevinrizza commented:
/lgtm

MartinBasti commented on operatorcourier/errors.py:
```diff
@@ -0,0 +1,63 @@
+
+class OpCourierQuayError(OpCourierError):
+    """An error occurred while pushing bundle to Quay.io"""
+    def __init__(self, msg, code):
```
Code is not supposed to be here, connection error has no HTTP code for example.

MartinBasti commented on operatorcourier/push.py:
```diff
@@ -42,15 +47,33 @@ def _push_to_registry(self, namespace, repository, release, bundle, auth_token):
+        logger.error(r.text)
+        msg = "Failed to push to app registry: {}"
+
+        try:
```
You may not even get here, IIRC Connection errors are raise directly by `requests.post()`, only HTTP status from reponse is raised by `raise_for_status`

chmeliik replied:
Would it then be safe to assume that if `requests.post()` doesn't fail, the response will always contain valid JSON content from  Quay? Or should I keep using `OpCourierQuayCommunicationError` for that and use `OpCourierQuayError` for exceptions raised by `requests.post()`?

kevinrizza replied:
The API unfortunately does not have a public spec and the source code is not yet open. However, I believe it is a fair assumption that, if requests.post() on that endpoint returns some kind of success that there will always be a json response, even if that response is empty. I haven't seen any behavior otherwise.

MartinBasti replied:
It needs something like this:
```python
try:
    r = request.post(.....)
except RequestsException as e:  # or Exception
   raise OpCourierQuayCommunicationError(str(e))
else:
   if r.status_code != 200:
     try:
        r_json = r.json()
     except Exception:
        r_json = {}
     raise OpCourierQuayErrorResponse('something', code, r_json)
```

chmeliik replied:
That looks reasonable. The only potential problem could that raising `OpCourierQuayCommunicationError` instead of `RequestException` would break backward compatibility. Of course, `OpCourierQuayError` could inherit from `RequestException` as well, though that would still break code like this: 
```python
try:
    api.build_verify_and_push()
except ValueError:
    # OpCourierQuayError gets caught here because it inherits from both
except RequestException:
    # and this doesn't happen
```
I'm probably making things unnecessarily complicated, but I'd still like to get your thoughts on it @MartinBasti  @kevinrizza

kevinrizza replied:
@chmeliik @MartinBasti Personally I think anything where `r.status_code != 200` should except one error type that inherits from `ValueError` and anything where `requests.post()` throws an exception itself should except another error type that inherits `RequestException`. To me I thought that was the whole point of separating these error types.

MartinBasti replied:
Ok, it could be done in this way

chmeliik commented:
Updated PR. Had to change the exception hierarchy a bit to allow for an exception that would inherit from the base exception but not from `ValueError`. Added a couple simple tests to make sure I didn't mess it up.
cc @MartinBasti @kevinrizza

kevinrizza commented:
/lgtm

chmeliik commented:
Rebased on master
